cpp/src/io/fst/agent_dfa.cuh

```diff
@@ -0,0 +1,672 @@
+ * transducer table are assignable.
+ * @tparam TransducedIndexOutItT A Random-access output iterator type to which indexes are written.
+ */
+template <typename TransducerTableT, typename TransducedOutItT, typename TransducedIndexOutItT>
+class DFASimulationCallbackWrapper {
+ public:
+  __host__ __device__ __forceinline__ DFASimulationCallbackWrapper(
+    TransducerTableT transducer_table, TransducedOutItT out_it, TransducedIndexOutItT out_idx_it)
+    : transducer_table(transducer_table), out_it(out_it), out_idx_it(out_idx_it), write(false)
+  {
+  }
+
+  template <typename OffsetT>
+  __host__ __device__ __forceinline__ void Init(OffsetT const& offset)
+  {
+    this->offset = offset;
+    if (!write) out_count = 0;
+  }
+
+  template <typename CharIndexT, typename StateIndexT, typename SymbolIndexT>
+  __host__ __device__ __forceinline__ void ReadSymbol(CharIndexT const character_index,
+                                                      StateIndexT const old_state,
+                                                      StateIndexT const new_state,
+                                                      SymbolIndexT const symbol_id)
+  {
+    uint32_t const count = transducer_table(old_state, symbol_id);
+    if (write) {
+      for (uint32_t out_char = 0; out_char < count; out_char++) {
+        out_it[out_count + out_char]     = transducer_table(old_state, symbol_id, out_char);
+        out_idx_it[out_count + out_char] = offset + character_index;
+      }
+    }
+    out_count += count;
+  }
+
+  __host__ __device__ __forceinline__ void TearDown() {}
+
+ public:
+  TransducerTableT const transducer_table;
+  TransducedOutItT out_it;
+  TransducedIndexOutItT out_idx_it;
+  uint32_t out_count;
+  uint32_t offset;
+  bool write;
+};
+
+/**
+ * @brief Helper class that transitions the state of multiple DFA instances simultaneously whenever
+ * a symbol is read.
+ *
+ * @tparam NUM_INSTANCES The number of DFA instances to keep track of
+ * @tparam TransitionTableT The transition table type used for looking up the new state for a
+ * current_state and a read_symbol.
+ */
+template <int32_t NUM_INSTANCES, typename TransitionTableT>
+class StateVectorTransitionOp {
+ public:
+  __host__ __device__ __forceinline__ StateVectorTransitionOp(
+    TransitionTableT const& transition_table, std::array<StateIndexT, NUM_INSTANCES>& state_vector)
+    : transition_table(transition_table), state_vector(state_vector)
+  {
+  }
+
+  template <typename CharIndexT, typename SymbolIndexT>
+  __host__ __device__ __forceinline__ void ReadSymbol(CharIndexT const& character_index,
+                                                      SymbolIndexT const read_symbol_id) const
+  {
+    for (int32_t i = 0; i < NUM_INSTANCES; ++i) {
+      state_vector[i] = transition_table(state_vector[i], read_symbol_id);
+    }
+  }
+
+ public:
+  std::array<StateIndexT, NUM_INSTANCES>& state_vector;
+  TransitionTableT const& transition_table;
```

Comment:
nit: I am always a bit wary of reference members, but with one of the members being const, at least operator= is deleted.

Reply:
Yup, I'm with you. But it's also not uncommon in CUB. E.g., the `Block*`-level algorithms keep a ref to their `TempStorage`.